File: api/app/services/credits.py
import time
from datetime import datetime, timedelta, timezone

from sqlalchemy.exc import OperationalError
from sqlalchemy.orm import Session

from app.models import CreditAccount, Subscription
from app.settings import settings


def _get_or_create_account(db: Session, user_key: str) -> CreditAccount:
    account = db.query(CreditAccount).filter(CreditAccount.user_key == user_key).first()
    if account:
        _maybe_reset_monthly(db, account)
        return account

    initial_balance = settings.FREE_MONTHLY_TOKENS if is_subscribed(db, user_key) else settings.FREE_SIGNUP_TOKENS
    account = CreditAccount(
        user_key=user_key,
        balance=initial_balance,
        last_reset_at=datetime.now(timezone.utc),
    )
    db.add(account)
    for attempt in range(3):
        try:
            db.commit()
            break
        except OperationalError:
            db.rollback()
            if attempt == 2:
                raise
            time.sleep(0.5 * (attempt + 1))
            db.add(account)
    db.refresh(account)
    return account


def _maybe_reset_monthly(db: Session, account: CreditAccount) -> None:
    """Grant the monthly token allowance if 30+ days since last reset.

    Only Pro (subscribed) users receive the monthly token top-up.
    """
    now = datetime.now(timezone.utc)
    last = account.last_reset_at
    subscribed = is_subscribed(db, account.user_key)
    if last is None:
        # Legacy account — seed it now, but only for subscribers.
        if subscribed:
            account.balance = max(account.balance, settings.FREE_MONTHLY_TOKENS)
        account.last_reset_at = now
        db.commit()
        return
    if last.tzinfo is None:
        last = last.replace(tzinfo=timezone.utc)
    if now - last >= timedelta(days=30):
        if subscribed:
            account.balance = max(account.balance, settings.FREE_MONTHLY_TOKENS)
        account.last_reset_at = now
        db.commit()


def is_subscribed(db: Session, user_key: str) -> bool:
    """Return True if the user has an active paid subscription."""
    return (
        db.query(Subscription)
        .filter(Subscription.user_key == user_key, Subscription.status == "active")
        .first()
    ) is not None


def get_credit_balance(db: Session, user_key: str) -> int:
    return _get_or_create_account(db, user_key).balance


def add_credits(db: Session, user_key: str, amount: int = 1) -> int:
    if amount < 0:
        raise ValueError("amount must be non-negative")

    account = _get_or_create_account(db, user_key)
    account.balance += amount
    db.commit()
    db.refresh(account)
    return account.balance


def award_job_post_credit(db: Session, user_key: str) -> dict:
    """Grant +1 free token for posting a job, capped per 30-day window.

    Replaces the old unlimited "post a job → get a token" loop. The cap
    (FREE_JOB_POST_TOKENS_PER_MONTH) is shared across web + WhatsApp because it
    keys off the same credit account. Returns:
        {"granted": bool, "balance": int, "remaining": int}
    where ``remaining`` is how many more free tokens can still be earned this window.
    """
    cap = settings.FREE_JOB_POST_TOKENS_PER_MONTH
    account = _get_or_create_account(db, user_key)

    now = datetime.now(timezone.utc)
    start = account.job_post_window_start
    if start is not None and start.tzinfo is None:
        start = start.replace(tzinfo=timezone.utc)
    if start is None or now - start >= timedelta(days=30):
        account.job_post_tokens_earned = 0
        account.job_post_window_start = now

    if account.job_post_tokens_earned >= cap:
        db.commit()
        return {"granted": False, "balance": account.balance, "remaining": 0}

    account.balance += 1
    account.job_post_tokens_earned += 1
    db.commit()
    db.refresh(account)
    return {
        "granted": True,
        "balance": account.balance,
        "remaining": max(0, cap - account.job_post_tokens_earned),
    }


def spend_credits(db: Session, user_key: str, amount: int = 1) -> int | None:
    if amount < 0:
        raise ValueError("amount must be non-negative")

    for attempt in range(3):
        try:
            account = _get_or_create_account(db, user_key)
            if account.balance < amount:
                return None
            account.balance -= amount
            db.commit()
            db.refresh(account)
            return account.balance
        except OperationalError:
            db.rollback()
            if attempt == 2:
                raise
            time.sleep(0.5 * (attempt + 1))

File: jobcarver-go/internal/credits/credits.go
// Package credits ports api/app/services/credits.py: the token-economy account
// logic (balances, monthly free grants, job-post earn cap, spending).
//
// Function names mirror the Python service so handlers read the same:
//
//	GetBalance -> get_credit_balance, AddCredits -> add_credits,
//	SpendCredits -> spend_credits, AwardJobPostCredit -> award_job_post_credit,
//	IsSubscribed -> is_subscribed.
package credits

import (
	"errors"
	"time"

	"gorm.io/gorm"

	"jobcarver/internal/config"
	"jobcarver/internal/models"
)

// ErrInsufficientCredits is returned by Spend when the balance is too low.
var ErrInsufficientCredits = errors.New("insufficient credits")

// IsSubscribed reports whether the user has an active paid subscription
// (ports is_subscribed).
func IsSubscribed(db *gorm.DB, userKey string) bool {
	var count int64
	db.Model(&models.Subscription{}).
		Where("user_key = ? AND status = ?", userKey, "active").
		Count(&count)
	return count > 0
}

// getOrCreateAccount ports _get_or_create_account: returns the user's account,
// creating it (initial balance = FreeMonthlyTokens if subscribed else
// FreeSignupTokens) if absent, and applying the monthly reset on existing ones.
// TODO(go-port): the Python retry-on-locked-DB loop is dropped — gorm + WAL
// busy_timeout handles contention.
func getOrCreateAccount(db *gorm.DB, userKey string) (*models.CreditAccount, error) {
	var account models.CreditAccount
	err := db.Where("user_key = ?", userKey).First(&account).Error
	if err == nil {
		maybeResetMonthly(db, &account)
		return &account, nil
	}
	if !errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, err
	}

	initial := config.S.FreeSignupTokens
	if IsSubscribed(db, userKey) {
		initial = config.S.FreeMonthlyTokens
	}
	now := time.Now().UTC()
	account = models.CreditAccount{
		UserKey:     userKey,
		Balance:     initial,
		LastResetAt: &now,
	}
	if err := db.Create(&account).Error; err != nil {
		return nil, err
	}
	return &account, nil
}

// maybeResetMonthly ports _maybe_reset_monthly: grant the monthly token
// allowance if 30+ days since last reset. Only subscribers receive the top-up.
func maybeResetMonthly(db *gorm.DB, account *models.CreditAccount) {
	now := time.Now().UTC()
	subscribed := IsSubscribed(db, account.UserKey)

	if account.LastResetAt == nil {
		if subscribed && account.Balance < config.S.FreeMonthlyTokens {
			account.Balance = config.S.FreeMonthlyTokens
		}
		account.LastResetAt = &now
		db.Save(account)
		return
	}
	if now.Sub(account.LastResetAt.UTC()) >= 30*24*time.Hour {
		if subscribed && account.Balance < config.S.FreeMonthlyTokens {
			account.Balance = config.S.FreeMonthlyTokens
		}
		account.LastResetAt = &now
		db.Save(account)
	}
}

// GetBalance returns the user's current token balance, creating the account if
// needed (ports get_credit_balance). Returns 0 on error.
func GetBalance(db *gorm.DB, userKey string) int {
	account, err := getOrCreateAccount(db, userKey)
	if err != nil {
		return 0
	}
	return account.Balance
}

// AddCredits adds amount tokens and returns the new balance (ports add_credits).
func AddCredits(db *gorm.DB, userKey string, amount int) (int, error) {
	if amount < 0 {
		return 0, errors.New("amount must be non-negative")
	}
	account, err := getOrCreateAccount(db, userKey)
	if err != nil {
		return 0, err
	}
	account.Balance += amount
	if err := db.Save(account).Error; err != nil {
		return 0, err
	}
	return account.Balance, nil
}

// SpendCredits deducts amount tokens. ok is false (and balance is the unchanged
// current balance) when the user has insufficient credits (ports spend_credits,
// where the Python returns None on insufficient funds).
func SpendCredits(db *gorm.DB, userKey string, amount int) (balance int, ok bool) {
	if amount < 0 {
		return 0, false
	}
	account, err := getOrCreateAccount(db, userKey)
	if err != nil {
		return 0, false
	}
	if account.Balance < amount {
		return account.Balance, false
	}
	account.Balance -= amount
	if err := db.Save(account).Error; err != nil {
		return account.Balance, false
	}
	return account.Balance, true
}

// AwardJobPostCredit grants +1 free token for posting a job, capped per 30-day
// window (FreeJobPostTokensPerMonth, shared web + WhatsApp). Ports
// award_job_post_credit. Returns (granted, balance, remaining).
func AwardJobPostCredit(db *gorm.DB, userKey string) (granted bool, balance int, remaining int) {
	cap := config.S.FreeJobPostTokensPerMonth
	account, err := getOrCreateAccount(db, userKey)
	if err != nil {
		return false, 0, 0
	}
	now := time.Now().UTC()
	if account.JobPostWindowStart == nil || now.Sub(account.JobPostWindowStart.UTC()) >= 30*24*time.Hour {
		account.JobPostTokensEarned = 0
		account.JobPostWindowStart = &now
	}
	if account.JobPostTokensEarned >= cap {
		db.Save(account)
		return false, account.Balance, 0
	}
	account.Balance++
	account.JobPostTokensEarned++
	db.Save(account)
	rem := cap - account.JobPostTokensEarned
	if rem < 0 {
		rem = 0
	}
	return true, account.Balance, rem
}

// ── Backward-compatible aliases ─────────────────────────────────────────────
// Kept so any handler written against the earlier draft contract still compiles.

// EnsureAccount creates the user's credit account if it does not yet exist.
func EnsureAccount(db *gorm.DB, userKey string) error {
	_, err := getOrCreateAccount(db, userKey)
	return err
}

// Grant is an alias for AddCredits that discards the new balance.
func Grant(db *gorm.DB, userKey string, amount int) error {
	_, err := AddCredits(db, userKey, amount)
	return err
}

// Spend is an alias for SpendCredits returning ErrInsufficientCredits when funds
// are too low.
func Spend(db *gorm.DB, userKey string, amount int) error {
	if _, ok := SpendCredits(db, userKey, amount); !ok {
		return ErrInsufficientCredits
	}
	return nil
}

// ApplyFreeMonthlyGrantIfDue applies the 30-day monthly reset for a user.
func ApplyFreeMonthlyGrantIfDue(db *gorm.DB, userKey string) {
	_, _ = getOrCreateAccount(db, userKey)
}
